text/0000-bee-model-crate/0000-bee-model-crate.md

```diff
@@ -0,0 +1,334 @@
+
+## Exposed Interface
+
+### Transaction
+
+ A transaction consists of following fields:
+
+```rust
+struct Transaction {
+
+    signature_fragments: String,
+    address: String,
+    value: i64,
+    obsolete_tag: String,
+    timestamp: i64,
+    current_index: usize,
+    last_index: usize,
+    bundle_hash: String,
+    trunk: String,
+    branch: String,
+    nonce: String,
+    tag: String,
+    attachment_timestamp: i64,
+    attachment_timestamp_lower_bound: i64,
+    attachment_timestamp_upper_bound: i64,
+
+}
+```
+
+As mentioned, Transactions are final. Transaction fields therefore should be only accessible by getter functions.
+Besides that, a transaction can only be created from a Transaction Builder, or from a constructor function which expects encoded bytes as received e.g. from a network socket.
+
+```rust
+impl Transaction {
+    
+    pub fn from_bytes(bytes: &Vec<u8>) -> Result<Transaction, TransactionBuilderValidationError> {
+        unimplemented!()
+    }
+
+    pub fn signature_fragments(&self) -> &String {
+        &self.signature_fragments
+    }
+
+    pub fn address(&self) -> &String {
+        &self.address
+    }
+
+    pub fn value(&self) -> &i64 {
+        &self.value
+    }
+
+    pub fn obsolete_tag(&self) -> &String {
+        &self.obsolete_tag
+    }
+
+    pub fn timestamp(&self) -> &i64 {
+        &self.timestamp
+    }
+
+    pub fn current_index(&self) -> &usize {
+        &self.current_index
+    }
+
+    pub fn last_index(&self) -> &usize {
+        &self.last_index
+    }
+
+    pub fn bundle_hash(&self) -> &String {
+        &self.bundle_hash
+    }
+
+    pub fn trunk(&self) -> &String {
+        &self.trunk
+    }
+
+    pub fn branch(&self) -> &String {
+        &self.branch
+    }
+
+    pub fn nonce(&self) -> &String {
+        &self.nonce
+    }
+
+    pub fn tag(&self) -> &String {
+        &self.tag
+    }
+
+    pub fn attachment_timestamp(&self) -> &i64 {
+        &self.attachment_timestamp
+    }
+
+    pub fn attachment_timestamp_lower_bound(&self) -> &i64 {
+        &self.attachment_timestamp_lower_bound
+    }
+
+    pub fn attachment_timestamp_upper_bound(&self) -> &i64 {
+        &self.attachment_timestamp_upper_bound
+    }
+
+}
+```
+
+The Transcation_Builder struct contains public accessible fields. 
+The set values can be validated in the build() function which then returns the constructed transaction.
+Moreover, also the Proof-Of-Work will be called in the build() function.
+
+```rust
+#[derive(Default)]
+pub struct TransactionBuilder {
+    pub signature_fragments: String,
+    pub address: String,
+    pub value: i64,
+    pub obsolete_tag: String,
+    pub current_index: usize,
+    pub last_index: usize,
+    pub bundle_hash: String,
+    pub trunk: String,
+    pub branch: String,
+    pub tag: String,
+    pub attachment_timestamp: i64,
+    pub attachment_timestamp_lower_bound: i64,
+    pub attachment_timestamp_upper_bound: i64,
+}
```

Comment:
For a transaction builder, the proper semantics are to keep the fields as `Option<T>`. Also, I would also leave the fields `non-pub` and only allow setting them through setter methods.

Reply:
> For a transaction builder, the proper semantics are to keep the fields as Option<T>
Great idea
> Also, I would also leave the fields non-pub and only allow setting them through setter methods.
Where would you put the validation of the set values? Directly in the setter, or as it currently is in `build()`? Currently the validation happens one time, else it could be done in each setter directly. Not sure what suites best here.

Reply:
Regarding `non-pub` there is the idea to build the actual Transaction (creating the struct) in Pow::compute(). Currently we have the Transaction fields `non-pub`. It seems like this way the struct can't get initialized.

Reply:
Let's use visibility we introduced yesterday then. I think `pub(crate)` should be fine.